File: CS/XpoImport/XpoImportHelper.cs
using DevExpress.Xpo;
using DevExpress.Xpo.Helpers;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace DevExpress.Sample {

    public abstract class XpoImportHelper {
        private IDataLayer dataLayer;
        private ICommandChannel commandChannel;
        private int batchSize;
        public XpoImportHelper(IDataLayer dataLayer, int batchSize) {
            this.dataLayer = dataLayer;
            this.commandChannel = dataLayer as ICommandChannel;
            this.batchSize = batchSize;
        }
        protected abstract object CreatePersistentObject(Session session, object sourceObject);
        public void Import(IEnumerable sourceObjects) {
            if (commandChannel != null) {
                ImportInBatches(sourceObjects);
            } else {
                ImportAtOnce(sourceObjects);
            }
        }

        private void ImportAtOnce(IEnumerable sourceObjects) {
            using (UnitOfWork uow = new UnitOfWork(dataLayer)) {
                foreach (object current in sourceObjects) {
                    object obj = CreatePersistentObject(uow, current);
                    uow.Save(obj);
                }
                uow.CommitChanges();
            }
        }

        private void ImportInBatches(IEnumerable sourceObjects) {
            commandChannel.Do(CommandChannelHelper.Command_ExplicitBeginTransaction, null);
            UnitOfWork uow = null;
            try {
                int i = 0;
                IEnumerator enumerator = sourceObjects.GetEnumerator();
                bool more = enumerator.MoveNext();
                while (more) {
                    if (uow == null) {
                        uow = new UnitOfWork(dataLayer);
                    }
                    object obj = CreatePersistentObject(uow, enumerator.Current);
                    uow.Save(obj);
                    more = enumerator.MoveNext();
                    i++;
                    if (!more || i >= batchSize) {
                        uow.CommitChanges();
                        uow.Dispose();
                        uow = null;
                        i = 0;
                    }
                }
                commandChannel.Do(CommandChannelHelper.Command_ExplicitCommitTransaction, null);
            } catch (Exception ex) {
                if (uow != null) {
                    uow.Dispose();
                }
                commandChannel.Do(CommandChannelHelper.Command_ExplicitRollbackTransaction, null);
                throw ex;
            }
        }
    }

}

File: VB/XpoImport/XpoImportHelper.vb
Imports DevExpress.Xpo
Imports DevExpress.Xpo.Helpers
Imports System
Imports System.Collections

Namespace DevExpress.Sample

    Public MustInherit Class XpoImportHelper

        Private dataLayer As IDataLayer

        Private commandChannel As ICommandChannel

        Private batchSize As Integer

        Public Sub New(ByVal dataLayer As IDataLayer, ByVal batchSize As Integer)
            Me.dataLayer = dataLayer
            commandChannel = TryCast(dataLayer, ICommandChannel)
            Me.batchSize = batchSize
        End Sub

        Protected MustOverride Function CreatePersistentObject(ByVal session As Session, ByVal sourceObject As Object) As Object

        Public Sub Import(ByVal sourceObjects As IEnumerable)
            If commandChannel IsNot Nothing Then
                ImportInBatches(sourceObjects)
            Else
                ImportAtOnce(sourceObjects)
            End If
        End Sub

        Private Sub ImportAtOnce(ByVal sourceObjects As IEnumerable)
            Using uow As UnitOfWork = New UnitOfWork(dataLayer)
                For Each current As Object In sourceObjects
                    Dim obj As Object = CreatePersistentObject(uow, current)
                    uow.Save(obj)
                Next

                uow.CommitChanges()
            End Using
        End Sub

        Private Sub ImportInBatches(ByVal sourceObjects As IEnumerable)
            commandChannel.Do(CommandChannelHelper.Command_ExplicitBeginTransaction, Nothing)
            Dim uow As UnitOfWork = Nothing
            Try
                Dim i As Integer = 0
                Dim enumerator As IEnumerator = sourceObjects.GetEnumerator()
                Dim more As Boolean = enumerator.MoveNext()
                While more
                    If uow Is Nothing Then
                        uow = New UnitOfWork(dataLayer)
                    End If

                    Dim obj As Object = CreatePersistentObject(uow, enumerator.Current)
                    uow.Save(obj)
                    more = enumerator.MoveNext()
                    i += 1
                    If Not more OrElse i >= batchSize Then
                        uow.CommitChanges()
                        uow.Dispose()
                        uow = Nothing
                        i = 0
                    End If
                End While

                commandChannel.Do(CommandChannelHelper.Command_ExplicitCommitTransaction, Nothing)
            Catch ex As Exception
                If uow IsNot Nothing Then
                    uow.Dispose()
                End If

                commandChannel.Do(CommandChannelHelper.Command_ExplicitRollbackTransaction, Nothing)
                Throw ex
            End Try
        End Sub
    End Class
End Namespace
